pkg/erofs/erofs.go

```diff
@@ -690,87 +724,155 @@ func (i *Inode) blocks() uint64 {
 //
 // [ (metadata block) inode | optional fields | dirent M+2 | dirent M+3 | name M+2 | name M+3 | optional padding ]
 //
-// All directory entries are _strictly_ recorded in alphabetical order.
-func (i *Inode) IterDirents(cb func(name string, typ uint8, nid uint64) error) error {
+// refer: https://docs.kernel.org/filesystems/erofs.html#directories
+func (i *Inode) getDirentName(d *Dirent, block blockData, lastDirent bool) ([]byte, error) {
+	var nameLen uint32
+	if lastDirent {
+		nameLen = block.size - uint32(d.NameOff)
+	} else {
+		nameLen = uint32(direntAfter(d).NameOff - d.NameOff)
+	}
+	if uint32(d.NameOff)+nameLen > block.size || nameLen > MaxNameLen || nameLen == 0 {
+		log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+		return nil, linuxerr.EUCLEAN
+	}
+	name, err := i.image.BytesAt(block.base+uint64(d.NameOff), uint64(nameLen))
+	if err != nil {
+		return nil, err
+	}
+	if lastDirent {
+		// Optional padding may exist at the end of a block.
+		n := bytes.IndexByte(name, 0)
+		if n == 0 {
+			log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+			return nil, linuxerr.EUCLEAN
+		}
+		if n != -1 {
+			name = name[:n]
+		}
+	}
+	return name, nil
+}
+
+// getDirent0 returns a pointer to the first dirent in the given block of this inode.
+func (i *Inode) getDirent0(block blockData) (*Dirent, error) {
+	d0, err := i.image.direntAt(block.base)
+	if err != nil {
+		return nil, err
+	}
+	// Apart from the offset of the first filename, nameOff0 also indicates
+	// the total number of dirents in this block.
+	if d0.NameOff < DirentSize || uint32(d0.NameOff) >= block.size {
+		log.Warningf("Invalid nameOff0 %v at inode (nid=%v)", d0.NameOff, i.Nid())
+		return nil, linuxerr.EUCLEAN
+	}
+	return d0, nil
+}
+
+// Lookup looks up a child by the name. The child inode number will be returned on success.
+func (i *Inode) Lookup(name string) (uint64, error) {
 	if !i.IsDir() {
-		return linuxerr.ENOTDIR
+		return 0, linuxerr.ENOTDIR
 	}
 
-	blocks := i.blocks()
-	blockSize := i.image.BlockSize()
+	// Currently (Go 1.21), there is no safe and efficient way to do three-way
+	// string comparisons, so let's convert the string to a byte slice first.
+	nameBytes := gohacks.ImmutableBytesFromString(name)
 
-	start := i.dataOff
-	if blocks == 1 && i.idataOff != 0 {
-		start = i.idataOff
+	// In EROFS, all directory entries are _strictly_ recorded in alphabetical
+	// order. The lookup is done by directly performing binary search on the
+	// disk data similar to what Linux does in fs/erofs/namei.c:erofs_namei().
+	type blockInfo struct {
+		block      blockData
```

Comment:
this field should be named `data`, since `block` is a bit ambiguous (blockInfo vs blockData).

Reply:
Makes sense, `block` is ambiguous currently. I think keeping the name `block` might be preferable, as other variables of the `blockData` type are also named `block`. Perhaps we could consider renaming `blockInfo` to `targetInfo` instead?

Reply:
Maybe get rid of this struct and just have `targetBlock` and `targetNumDirents`?

Reply:
Good idea. Will do. Thanks!